File: Clippy/Services/AppCoordinator.swift
import Foundation
import SwiftUI

@MainActor
final class AppCoordinator: ObservableObject {
    static let shared = AppCoordinator()

    @Published var lastClip: Clip?
    @Published var showClipSavedBanner = false
    @Published var errorMessage: String?
    @Published var showOnboarding: Bool

    private let settings = AppSettings.shared
    private let recorder = ScreenRecorder.shared
    private let clips = ClipManager.shared
    private let voice = VoiceCommandListener.shared

    private init() {
        showOnboarding = !AppSettings.shared.hasCompletedOnboarding
    }

    func bootstrap() {
        HotkeyManager.shared.onTrigger = { [weak self] in
            Task { @MainActor in
                await self?.triggerClip(source: .hotkey)
            }
        }
        HotkeyManager.shared.register(binding: settings.hotkey)

        voice.onClipCommand = { [weak self] in
            ClippyDebugLog.shared.log("Clip", "Voice command received")
            Task { @MainActor in
                await self?.triggerClip(source: .voice)
            }
        }

        voice.onOnboardingClipCommand = { [weak self] in
            Task { @MainActor in
                self?.completeOnboarding(fromVoiceDemo: true)
            }
        }

        if settings.hasCompletedOnboarding {
            startBackgroundServices()
        } else {
            recorder.requestScreenCaptureAccess()
        }
    }

    func startBackgroundServices() {
        recorder.requestScreenCaptureAccess()

        Task {
            await DisplayStore.shared.refreshDisplays()
            await recorder.startCapture()
            if settings.voiceCommandsEnabled {
                try? await Task.sleep(nanoseconds: 2_000_000_000)
                await voice.prepareAndStart()
            }
        }
    }

    func applyOnboardingAudioDevices() {
        if !settings.preferredAudioOutputUID.isEmpty {
            AudioDeviceManager.setSystemDefaultOutputDevice(uid: settings.preferredAudioOutputUID)
        }
        if !settings.preferredMicrophoneUID.isEmpty {
            AudioDeviceManager.setSystemDefaultInputDevice(uid: settings.preferredMicrophoneUID)
        }
        voice.refreshMicrophone()
    }

    func beginOnboardingVoicePractice() {
        applyOnboardingAudioDevices()
        settings.voiceCommandsEnabled = true

        Task {
            await DisplayStore.shared.refreshDisplays()
            await recorder.startCapture()
            try? await Task.sleep(nanoseconds: 1_000_000_000)
            voice.onboardingMode = true
            await voice.prepareAndStart()
        }
    }

    func completeOnboarding(fromVoiceDemo: Bool) {
        if fromVoiceDemo {
            SoundPlayer.shared.playClipSound()
        }
        voice.onboardingMode = false
        settings.hasCompletedOnboarding = true
        settings.voiceCommandsEnabled = true
        withAnimation(ClippyTheme.spring) {
            showOnboarding = false
        }
        Task {
            await voice.prepareAndStart()
        }
    }

    func refreshHotkey() {
        HotkeyManager.shared.register(binding: settings.hotkey)
    }

    func refreshVoiceListening() {
        if settings.voiceCommandsEnabled {
            Task {
                try? await Task.sleep(nanoseconds: 500_000_000)
                await voice.prepareAndStart()
            }
        } else {
            Task { await voice.stopListening() }
        }
    }

    func refreshMicrophone() {
        voice.refreshMicrophone()
    }

    func refreshDisplay() {
        Task {
            await DisplayStore.shared.refreshDisplays()
            await recorder.restartCapture()
        }
    }

    func refreshCaptureQuality() {
        Task {
            await recorder.restartCapture()
        }
    }

    enum ClipSource {
        case hotkey
        case voice
        case button
    }

    func triggerClip(source: ClipSource) async {
        guard !recorder.isClipping else { return }
        let requiredSeconds = settings.clipDuration.seconds
        guard recorder.isBufferReady else {
            let msg = "Buffer is still filling — wait until status shows Ready."
            ClippyDebugLog.shared.log("Clip", "Blocked — buffer not ready. \(RecorderDiagnostics.snapshot(recorder: recorder))")
            errorMessage = msg
            return
        }
        guard recorder.bufferedSeconds >= requiredSeconds - 0.5 else {
            let msg = "Only \(Int(recorder.bufferedSeconds))s buffered — wait for \(Int(requiredSeconds))s before clipping."
            ClippyDebugLog.shared.log("Clip", "Blocked — insufficient buffer: \(recorder.bufferedSeconds)s / \(requiredSeconds)s")
            errorMessage = msg
            return
        }

        ClippyDebugLog.shared.log("Clip", "Triggered via \(source)")
        SoundPlayer.shared.playClipSound()

        do {
            let result = try await recorder.createClip(maxDuration: settings.clipDuration.seconds)
            let clip = try await clips.addClip(from: result.url, duration: result.duration)
            try? FileManager.default.removeItem(at: result.url)
            lastClip = clip
            ClippyDebugLog.shared.log("Clip", "Saved clip \(clip.fileName) duration=\(result.duration)s")
            withAnimation(ClippyTheme.springBouncy) {
                showClipSavedBanner = true
            }
            DispatchQueue.main.asyncAfter(deadline: .now() + 2.5) {
                withAnimation(ClippyTheme.easeOut) {
                    self.showClipSavedBanner = false
                }
            }
        } catch {
            ClippyDebugLog.shared.logError("Clip", error, context: "triggerClip(\(source))")
            if !recorder.lastClipDebugSummary.isEmpty {
                ClippyDebugLog.shared.log("Clip", recorder.lastClipDebugSummary)
            }
            VoiceDiagnostics.logSnapshot(voice: voice)
            errorMessage = error.localizedDescription
        }
    }
}

File: windows/Clippy/Services/AppCoordinator.cs
using Clippy.Models;

namespace Clippy.Services;

public sealed class AppCoordinator
{
    private static AppCoordinator? _instance;
    public static AppCoordinator Instance => _instance ??= new AppCoordinator();

    public bool ShowOnboarding { get; private set; } = !AppSettings.Instance.HasCompletedOnboarding;
    public bool ShowClipSavedBanner { get; private set; }
    public string? ErrorMessage { get; private set; }
    public Clip? LastClip { get; private set; }

    public event Action? StateChanged;

    private AppCoordinator()
    {
    }

    public void Bootstrap(nint windowHandle)
    {
        HotkeyManager.Instance.AttachWindow(windowHandle);
        HotkeyManager.Instance.OnTrigger = () => _ = TriggerClipAsync(ClipSource.Hotkey);
        HotkeyManager.Instance.Register(AppSettings.Instance.Hotkey);

        VoiceCommandListener.Instance.OnClipCommand = () => _ = TriggerClipAsync(ClipSource.Voice);
        VoiceCommandListener.Instance.OnOnboardingClipCommand = () => CompleteOnboarding(fromVoiceDemo: true);

        if (AppSettings.Instance.HasCompletedOnboarding)
        {
            StartBackgroundServices();
        }
        else
        {
            ScreenRecorder.Instance.RequestScreenCaptureAccess();
        }
    }

    public void StartBackgroundServices()
    {
        ScreenRecorder.Instance.RequestScreenCaptureAccess();
        _ = Task.Run(async () =>
        {
            await ScreenRecorder.Instance.StartCaptureAsync();
            if (AppSettings.Instance.VoiceCommandsEnabled)
            {
                await Task.Delay(2000);
                await VoiceCommandListener.Instance.PrepareAndStartAsync();
            }
        });
    }

    public void ApplyOnboardingAudioDevices()
    {
        var settings = AppSettings.Instance;
        if (!string.IsNullOrEmpty(settings.PreferredAudioOutputId))
        {
            AudioDeviceManager.SetDefaultOutputDevice(settings.PreferredAudioOutputId);
        }

        if (!string.IsNullOrEmpty(settings.PreferredMicrophoneId))
        {
            AudioDeviceManager.SetDefaultInputDevice(settings.PreferredMicrophoneId);
        }

        VoiceCommandListener.Instance.RefreshMicrophone();
    }

    public void BeginOnboardingVoicePractice()
    {
        ApplyOnboardingAudioDevices();
        AppSettings.Instance.VoiceCommandsEnabled = true;
        AppSettings.Instance.Persist();

        VoiceCommandListener.Instance.OnboardingMode = true;
        _ = Task.Run(async () =>
        {
            await ScreenRecorder.Instance.StartCaptureAsync();
            await Task.Delay(1000);
            await VoiceCommandListener.Instance.PrepareAndStartAsync();
        });
    }

    public void CompleteOnboarding(bool fromVoiceDemo)
    {
        if (fromVoiceDemo)
        {
            SoundPlayer.Instance.PlayClipSound();
        }

        VoiceCommandListener.Instance.OnboardingMode = false;
        AppSettings.Instance.HasCompletedOnboarding = true;
        AppSettings.Instance.VoiceCommandsEnabled = true;
        AppSettings.Instance.Persist();
        ShowOnboarding = false;
        NotifyStateChanged();
        StartBackgroundServices();

        _ = VoiceCommandListener.Instance.PrepareAndStartAsync();
    }

    public void RefreshHotkey() =>
        HotkeyManager.Instance.Register(AppSettings.Instance.Hotkey);

    public void RefreshVoiceListening()
    {
        VoiceCommandListener.Instance.ResetSpeechBlock();
        if (AppSettings.Instance.VoiceCommandsEnabled)
        {
            _ = Task.Run(async () =>
            {
                await Task.Delay(500);
                await VoiceCommandListener.Instance.PrepareAndStartAsync();
            });
        }
        else
        {
            _ = VoiceCommandListener.Instance.StopListeningAsync();
        }
    }

    public void RefreshMicrophone() => VoiceCommandListener.Instance.RefreshMicrophone();

    public void RefreshDisplay() =>
        _ = ScreenRecorder.Instance.RestartCaptureAsync();

    public void RefreshCaptureQuality() =>
        _ = ScreenRecorder.Instance.RestartCaptureAsync();

    public enum ClipSource
    {
        Hotkey,
        Voice,
        Button
    }

    public async Task TriggerClipAsync(ClipSource source)
    {
        var recorder = ScreenRecorder.Instance;
        if (recorder.IsClipping) return;

        var required = AppSettings.Instance.ClipDuration.Seconds();
        if (!recorder.IsBufferReady)
        {
            ErrorMessage = "Buffer is still filling — wait until status shows Ready.";
            ClippyDebugLog.Instance.Log("Clip",
                $"Blocked — buffer not ready. {RecorderDiagnostics.Snapshot(recorder)}");
            NotifyStateChanged();
            return;
        }

        if (recorder.BufferedSeconds < required - 0.5)
        {
            ErrorMessage =
                $"Only {(int)recorder.BufferedSeconds}s buffered — wait for {(int)required}s before clipping.";
            ClippyDebugLog.Instance.Log("Clip",
                $"Blocked — insufficient buffer: {recorder.BufferedSeconds}s / {required}s");
            NotifyStateChanged();
            return;
        }

        ClippyDebugLog.Instance.Log("Clip", $"Triggered via {source}");
        SoundPlayer.Instance.PlayClipSound();

        try
        {
            var result = await recorder.CreateClipAsync(required);
            var clip = await ClipManager.Instance.AddClipAsync(result.Path, result.Duration);
            try { File.Delete(result.Path); } catch { }

            LastClip = clip;
            ClippyDebugLog.Instance.Log("Clip", $"Saved clip {clip.FileName} duration={result.Duration}s");
            ShowClipSavedBanner = true;
            NotifyStateChanged();

            _ = Task.Run(async () =>
            {
                await Task.Delay(2500);
                ShowClipSavedBanner = false;
                NotifyStateChanged();
            });
        }
        catch (Exception ex)
        {
            ClippyDebugLog.Instance.LogError("Clip", ex, $"triggerClip({source})");
            if (!string.IsNullOrEmpty(recorder.LastClipDebugSummary))
            {
                ClippyDebugLog.Instance.Log("Clip", recorder.LastClipDebugSummary);
            }

            VoiceDiagnostics.LogSnapshot(VoiceCommandListener.Instance);
            ErrorMessage = ex.Message;
            NotifyStateChanged();
        }
    }

    public void ClearError()
    {
        ErrorMessage = null;
        NotifyStateChanged();
    }

    private void NotifyStateChanged() => StateChanged?.Invoke();
}
